contracts/YieldCrabLever.sol

```diff
@@ -0,0 +1,451 @@
+        ICrabStrategy(0x3B960E47784150F5a63777201ee2B15253D713e8);
+
+    ISwapRouter public constant swapRouter =
+        ISwapRouter(0xE592427A0AEce92De3Edee1F18E0157C05861564);
+    IERC20 public crab;
+    bytes6 public constant crabId = 0x333800000000;
+    bytes6 public constant wethId = 0x303000000000;
+    bytes6 public constant daiId = 0x303100000000;
+    bytes6 public constant usdcId = 0x303200000000;
+
+    constructor(Giver giver_) YieldLeverBase(giver_) {}
+
+    /// @notice Invest by creating a levered vault.
+    /// The steps are as follows:
+    /// 1. Based on the ilk transfer it from the user
+    /// @param seriesId The series to invest in
+    /// @param ilkId The collateral to use
+    /// @param amountToInvest The amount of ETH/WETH/DAI/USDC supplied by the user
+    /// @param borrowAmount The amount of fyToken to be borrowed
+    /// @param minCollateral to be received
+    function invest(
+        bytes6 seriesId,
+        bytes6 ilkId,
+        uint256 amountToInvest,
+        uint256 borrowAmount,
+        uint256 minCollateral
+    ) external payable returns (bytes12 vaultId) {
+        IPool pool = IPool(ladle.pools(seriesId));
+        IMaturingToken fyToken = pool.fyToken();
+        // Depend on ilkId we will have to choose the operation
+        if (ilkId != wethId) {
+            // transfer the amountToInvest to this contract
+            pool.base().safeTransferFrom(
+                msg.sender,
+                address(this),
+                amountToInvest
+            );
+        } else {
+            if (msg.value == 0) {
+                // We are dealing with WETH so we need to transfer it from the
+                pool.base().safeTransferFrom(
+                    msg.sender,
+                    address(this),
+                    amountToInvest
+                );
+            }
+        }
+
+        // Build the vault
+        (vaultId, ) = ladle.build(seriesId, crabId, 0);
+
+        bytes memory data = bytes.concat(
+            bytes1(uint8(uint256(Operation.BORROW))), //[0]
+            seriesId, // [1:7]
+            vaultId, // [7:19]
+            ilkId, // [19:25]
+            bytes32(amountToInvest) // [25:57]
+        );
+
+        bool success = IERC3156FlashLender(address(fyToken)).flashLoan(
+            this, // Loan Receiver
+            address(fyToken), // Loan Token
+            borrowAmount, // Loan Amount
+            data
+        );
+
+        if (!success) revert FlashLoanFailure();
+
+        DataTypes.Balances memory balances = cauldron.balances(vaultId);
+
+        // This is the amount to deposit, so we check for slippage here. As
+        // long as we end up with the desired amount, it doesn't matter what
+        // slippage occurred where.
+        if (balances.ink < minCollateral) revert SlippageFailure();
+
+        giver.give(vaultId, msg.sender);
+
+        if (address(this).balance > 0)
+            payable(msg.sender).call{value: address(this).balance}("");
```

Comment:
Probably nice to leave a comment for the auditor that there is no risk of reentrancy here because this is the end of the execution.